controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -348,27 +348,49 @@ private boolean isUnhealthyInAnotherZone(Application application, JobId job) {
+    private void abortIfOutdated(DeploymentStatus status, Map<JobId, List<DeploymentStatus.Job>> jobs, JobId job) {
+        status.jobs().get(job)
+              .flatMap(JobStatus::lastTriggered)
+              .filter(last -> ! last.hasEnded())
+              .ifPresent(last -> {
+                  if (jobs.get(job).stream().noneMatch(versions ->    versions.versions().targetsMatch(last.versions())
+                                                                   && versions.versions().sourcesMatchIfPresent(last.versions()))) {
+                      log.log(Level.INFO, "Aborting outdated run " + last);
+                      controller.jobController().abort(last.id());
+                  }
+              });
+    }
 
-        Run last = status.lastTriggered().get();
-        if (versionsList.stream().noneMatch(versions ->    versions.targetsMatch(last.versions())
-                                                        && versions.sourcesMatchIfPresent(last.versions())))
-            controller.jobController().abort(last.id());
+    /** Returns whether the job is free to start, and also aborts it if it's running with outdated versions. */
+    private boolean abortIfRunning(DeploymentStatus status, Map<JobId, List<DeploymentStatus.Job>> jobs, JobId job) {
+        abortIfOutdated(status, jobs, job);
+        boolean blocked = status.jobs().get(job).get().isRunning();
+
+        if ( ! job.type().isTest()) {
+            Optional<JobStatus> productionTest = JobType.testFrom(controller.system(), job.type().zone(controller.system()).region())
+                                                        .map(type -> new JobId(job.application(), type))
+                                                        .flatMap(status.jobs()::get);
+            if (productionTest.isPresent()) {
+                abortIfOutdated(status, jobs, productionTest.get().id());
+                // Production deployments are also blocked by their declared tests, if the next versions to run
+                //  for those are not the same as the versions we're considering running in the deployment job now.
+                if (productionTest.map(JobStatus::id).map(jobs::get)
+                                  .map(versions -> ! versions.get(0).versions().targetsMatch(jobs.get(job).get(0).versions()))
+                                  .orElse(false))
+                    blocked = true;
+            }
+        }
 
-        return false;
+        return ! blocked;
     }
 
     // ---------- Change management o_O ----------
 
     private boolean acceptNewApplicationVersion(DeploymentStatus status, InstanceName instance) {
         if (status.application().deploymentSpec().instance(instance).isEmpty()) return false; // Unknown instance.
-        if (status.hasFailures()) return true; // Allow changes to fix upgrade or previous revision problems.
+        if (status.hasFailures(instance)) return true; // Allow changes to fix upgrade or previous revision problems.
```

Comment:
This should be refined even further, to not allow a revision if it is the new revision that is failing eager system/staging tests ... :P